### PR TITLE
App method tracing

### PR DESCRIPTION
**Pull request by johnno1962**
Hi @zenangst, a new feature for InjectionIII for you. It adds a “track app” menu bar item which turns on  tracing of all calls to class methods defined in the main app bundle. See what you think..

### CODE REVIEW

**Comment by zenangst** [comment]
@johnno1962 looks good to me! 😎

File: InjectionBundle/SwiftInjection.swift [hunk 1]
```diff
@@ -421,37 +421,4 @@ public struct ClassMetadataSwift {
...
-
-        if let demangledNamePtr = demangledNamePtr {
-            let demangledName = String(cString: demangledNamePtr)
-            free(demangledNamePtr)
-            return demangledName
-        }
-        return mangledName
-    }
-}
-#endif
```
[the remark this answers is not in the document]
Comment (johnno1962) [reply InjectionBundle/SwiftInjection.swift]: It's not referenced in the code and ambiguous against a version in the SwiftTrace source.

**Comment by johnno1962** [comment]
Did you try it out? Any problems with projects you have? Merge it if you think it would be useful.

**Comment by zenangst** [comment]
@johnno1962 only checked the source code, might have time to give it a go this evening :)

**Comment by johnno1962** [comment]
This PR also fixes “Xprobe” which was really broken.

**Comment by johnno1962** [comment]
Now traces exit as well as entry point and times method invocations.

**Comment by johnno1962** [comment]
```Swift
        Unwrap.LearnCoordinator.activeStudyReview.setter : Swift.Optional<Unwrap.StudyReview> 0.0ms
              Unwrap.TappableTextView.addCustomizations() -> () 0.2ms
            -[TappableTextView initWithFrame:textContainer:] -> @56@0:8{CGRect={CGPoint=dd}{CGSize=dd}}16@48 2.4ms
          -[StudyTextView initWithFrame:textContainer:] -> @56@0:8{CGRect={CGPoint=dd}{CGSize=dd}}16@48 2.4ms
        -[StudyViewController initWithNibName:bundle:] -> @32@0:8@16@24 2.5ms
        Unwrap.StudyViewController.chapter.setter : Swift.String 0.0ms
          Unwrap.StudyViewController.configureNavigation() -> () 0.0ms
        Unwrap.StudyViewController.coordinator.setter : Swift.Optional<Unwrap.LearnCoordinator> 0.1ms
      Unwrap.LearnCoordinator.studyViewController(for: Swift.String) -> Unwrap.StudyViewController 5.7ms
          -[CoordinatedNavigationController initWithNibName:bundle:] -> @32@0:8@16@24 0.0ms
        -[CoordinatedNavigationController initWithRootViewController:] -> @24@0:8@16 6.0ms
      Unwrap.LearnCoordinator.startStudying(using: __C.UIViewController) -> () 16.4ms
    Unwrap.LearnCoordinator.startStudying(title: Swift.String) -> () 22.1ms
  Unwrap.LearnViewController.startStudying(title: Swift.String) -> () 22.1ms
-[LearnDataSource tableView:didSelectRowAtIndexPath:] -> v32@0:8@16@24 22.2ms
```

**Comment by johnno1962** [comment]
Hi @zenangst, have you had a chance to look at this yet? It’s working pretty well and I’d like to get a release out to the App Store before too long. What are your thoughts?

**Comment by grav** [comment]
@johnno1962 I tried this branch and got it working in the end, but there were a couple of caveats:
1. I couldn't checkout `04f9725` of SwiftTrace:
```
$ git submodule update --init --recursive
error: Server does not allow request for unadvertised object 04f97251c18a5c4822825e4a9684955d66bf86a8
Fetched in submodule path 'SwiftTrace', but it did not contain 04f97251c18a5c4822825e4a9684955d66bf86a8. Direct fetching of that commit failed.
```
but `441204c` seemed to work fine.

2. I had to change some of the run scripts to use a different path for Xcode. It was hardcoded to /Applications/Xcode101.app/...`.

**Comment by johnno1962** [comment]
Apologies for that - a few too many --amend commits. Did the tracing work in your project?

**Comment by zenangst** [comment]
@johnno1962 I was unable to test the branch because `SwiftTrace.swift` is missing

**Comment by johnno1962** [comment]
You’ll be needing one of those. I don’t understand, you used: 

git clone https://github.com/johnno1962/InjectionIII -b app-tracing --recurse-submodules?

```
Johns-MacBook-Air:tmp johnholdsworth$ find InjectionIII/ -name 'SwiftTrace*'
InjectionIII//.git/modules/SwiftTrace
InjectionIII//SwiftTrace
InjectionIII//SwiftTrace/SwiftTrace
InjectionIII//SwiftTrace/SwiftTrace/SwiftTrace.h
InjectionIII//SwiftTrace/SwiftTrace/SwiftTrace.mm
InjectionIII//SwiftTrace/SwiftTrace/SwiftTrace.swift
InjectionIII//SwiftTrace/SwiftTrace.gif
InjectionIII//SwiftTrace/SwiftTrace.podspec
InjectionIII//SwiftTrace/SwiftTraceApp
InjectionIII//SwiftTrace/SwiftTraceApp.xcodeproj
InjectionIII//SwiftTrace/SwiftTraceOSX
InjectionIII//SwiftTrace/SwiftTraceTests
InjectionIII//SwiftTrace/SwiftTraceTests/SwiftTraceTests.swift
InjectionIII//SwiftTrace/SwiftTraceXTests
InjectionIII//SwiftTrace/SwiftTraceXTests/SwiftTraceXTests.swift
```

**Comment by zenangst** [comment]
Running into a different error when building now 🤔 

image

**Comment by johnno1962** [comment]
Yeesh, sorry about that… Mumble mumble Xcode versions.. Mumble mumble. Should be OK now.

**Comment by zenangst** [comment]
@johnno1962 it builds and works now. Tested it in a client project, works like intended! Good job @johnno1962

**Comment by zenangst** [comment]
Hmm I am seeing some crashes when I have tracing enabled.

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x000000010b3eacb4 libswiftCore.dylib`swift::gatherWrittenGenericArgs(swift::TargetMetadata<swift::InProcess> const*, swift::TargetTypeContextDescriptor<swift::InProcess> const*, std::__1::vector<swift::TargetMetadata<swift::InProcess> const*, std::__1::allocator<swift::TargetMetadata<swift::InProcess> const*> >&) + 20
    frame #1: 0x000000010b3cb64f libswiftCore.dylib`swift::_swift_buildDemanglingForMetadata(swift::TargetMetadata<swift::InProcess> const*, swift::Demangle::Demangler&) + 655
    frame #2: 0x000000010b3cb691 libswiftCore.dylib`swift::_swift_buildDemanglingForMetadata(swift::TargetMetadata<swift::InProcess> const*, swift::Demangle::Demangler&) + 721
    frame #3: 0x000000010b3da4db libswiftCore.dylib`_swift_initClassMetadataImpl(swift::TargetClassMetadata<swift::InProcess>*, swift::ClassLayoutFlags, unsigned long, swift::TypeLayout const* const*, unsigned long*, bool) + 2443
    frame #4: 0x000000010b3da8ef libswiftCore.dylib`swift_initClassMetadata2 + 15
    frame #5: 0x000000010576358c NRK-TV-tvOS`type metadata completion function for PageViewController at <compiler-generated>:0
    frame #6: 0x000000010b3e16fd libswiftCore.dylib`swift::MetadataCacheEntryBase<(anonymous namespace)::GenericCacheEntry, void const*>::doInitialization(swift::ConcurrencyControl&, swift::MetadataCompletionQueueEntry*, swift::MetadataRequest) + 285
    frame #7: 0x000000010b3d7f24 libswiftCore.dylib`swift_getGenericMetadata + 1476
    frame #8: 0x000000010575fe2c {TargetApp}`type metadata accessor for PageViewController at <compiler-generated>:0
```

**Comment by johnno1962** [comment]
Thanks fo for trying it out @zenangst. Are you combining it with injection? That might be unwise though it should work. Hope reliable is it on a scale from 1 to 10. Occasional crashes/frequent crashes?

**Comment by zenangst** [comment]
@johnno1962 I think this is good to go! 👍

**Comment by johnno1962** [comment]
Merge ahoy?

**Pull request merged**